Sources/ComposableArchitecture/Reducer/ReducerBuilder.swift

```diff
@@ -403,5 +329,138 @@ public typealias ReducerBuilderOf<R: ReducerProtocol> = ReducerBuilder<R.State,
         r9
       )
     }
+
+    @inlinable
+    public static func buildEither<R0: ReducerProtocol, R1: ReducerProtocol>(
+      first reducer: R0
+    ) -> _Conditional<R0, R1>
+    where R0.State == State, R0.Action == Action {
+      .first(reducer)
+    }
+
+    @inlinable
+    public static func buildEither<R0: ReducerProtocol, R1: ReducerProtocol>(
+      second reducer: R1
+    ) -> _Conditional<R0, R1>
+    where R1.State == State, R1.Action == Action {
+      .second(reducer)
+    }
+
+    @inlinable
+    public static func buildExpression<R: ReducerProtocol>(_ expression: R) -> R
+    where R.State == State, R.Action == Action {
+      expression
+    }
+
+    @inlinable
+    public static func buildFinalResult<R: ReducerProtocol>(_ reducer: R) -> R
+    where R.State == State, R.Action == Action {
+      reducer
+    }
+
+    @_disfavoredOverload
+    @inlinable
+    public static func buildFinalResult<R: ReducerProtocol>(_ reducer: R) -> Reduce<State, Action>
+    where R.State == State, R.Action == Action {
+      Reduce(reducer)
+    }
+
+    @inlinable
+    public static func buildLimitedAvailability<R: ReducerProtocol>(
+      _ wrapped: R
+    ) -> _Optional<R>
+    where R.State == State, R.Action == Action {
+      _Optional(wrapped: wrapped)
+    }
+
+    @inlinable
+    public static func buildOptional<R: ReducerProtocol>(_ wrapped: R?) -> _Optional<R>
```

Comment:
SwiftUI makes `Optional<some View>` a `View` itself. Was making `Optional<some ReducerProtocol>` a `ReducerProtocol` instead of using  this `_Optional<some Reducer>` considered at some point?

Reply:
For example, with
```swift
extension Optional: ReducerProtocol where Wrapped: ReducerProtocol {
  public func reduce(into state: inout Wrapped.State, action: Wrapped.Action)
    -> EffectTask<Wrapped.Action>
  {
    switch self {
    case .none:
      return .none
    case .some(let wrapped):
      return wrapped.reduce(into: &state, action: action)
    }
  }
}
```
We can return directly `wrapped` in `buildOptional`, and `Optional.some(wrapped)` in `buildLimitedAvailability`, while using opaque reducers.

BTW, I've just spotted this thing while checking in SwiftUI interface:
```swift
@available(iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0, *)
@inlinable @inline(__always) internal func asOptional<T>(_ value: T) -> T? {
    func unwrap<T>() -> T { value as! T }
    let optionalValue: T? = unwrap()
    return optionalValue
}
```
This is some weird gymnastic and I'm not getting why they don't use `Optional.some` instead. But this is mostly trivia and unrelated to the PR.

Reply:
@tgrapperon Good find! I'm not sure why we didn't conform optional directly, maybe because of a bug with type aliases in pre-5.7. Added support and removed `_Optional`, thanks!

We also realized our `buildLimitedAvailability` wasn't right. This function needs to erase the reducer somehow. I think we snuck a fix into this PR, though.

Reply:
> We also realized our buildLimitedAvailability wasn't right.

Indeed! I was also seeing it wrong.

Reply:
Do you think we still need the `buildOptional` block? As I'm seeing it, it should be handled by the `buildBlock(reducer: R?)` now that R? is a reducer, isn't it?

Reply:
Nevermind, it is required to work with `buildLimitedAvailability` at least.

Reply:
@tgrapperon Just merged and missed this. What hit `main` should at least be an improvement, but if you find other tweaks that can be made def PR!

Reply:
No, I think we need this. As I'm getting it now:
```swift
if #available(iOS 16, *) {
  R()
}
```
will `buildLimitedAvailability` `R` into `AnyReducer` and then `buildOptional` it to produce `AnyReducer?`. If you remove `buildOptional`, the test for `buildLimitedAvailability` doesn't pass anymore, so I guess this is what happens (and it is maybe involved with other `blocks`/situations).

Reply:
@stephencelis `buildOptional` is involved as soon as we have unbalanced `if` checks of some sort, so it is required for sure.